src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs

```diff
@@ -82,24 +97,41 @@ public CorsResult EvaluatePolicy(HttpContext context, CorsPolicy policy)
         public virtual void EvaluateRequest(HttpContext context, CorsPolicy policy, CorsResult result)
         {
             var origin = context.Request.Headers[CorsConstants.Origin];
-            if (StringValues.IsNullOrEmpty(origin) || !policy.AllowAnyOrigin && !policy.Origins.Contains(origin))
+            if (StringValues.IsNullOrEmpty(origin))
             {
                 return;
             }
 
+            if (!policy.AllowAnyOrigin && !policy.Origins.Contains(origin))
+            {
+                _logger?.RequestHasOriginHeader();
+                _logger.PolicyFailure($"Request origin {origin} does not have permission to access the resource.");
```

Comment:
Cors specification is strict when matching entries. I think its better to add a brief text about it in the messages, otherwise users tend to overlook the tiny details.